8267554: Support loading stylesheets from data-URIs

mstr2 opened the pull request:
This PR extends data URI support to allow stylesheets to be loaded from data URIs.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8267554](https://bugs.openjdk.java.net/browse/JDK-8267554): Support loading stylesheets from data-URIs


### Reviewers
 * [Kevin Rushforth](https://openjdk.java.net/census#kcr) (@kevinrushforth - **Reviewer**)
 * [Ajit Ghaisas](https://openjdk.java.net/census#aghaisas) (@aghaisas - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jfx pull/536/head:pull/536` \
`$ git checkout pull/536`

Update a local copy of the PR: \
`$ git checkout pull/536` \
`$ git pull https://git.openjdk.java.net/jfx pull/536/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 536`

View PR using the GUI difftool: \
`$ git pr show -t 536`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jfx/pull/536.diff">https://git.openjdk.java.net/jfx/pull/536.diff</a>

</details>

kevinrushforth commented:
/reviewers 2
/csr

kevinrushforth commented on modules/javafx.graphics/src/main/java/javafx/css/Stylesheet.java:
```diff
@@ -278,25 +279,46 @@ final void readBinary(int bssVersion, DataInputStream is, String[] strings)
+     * @throws IOException if the binary stream corresponds to a more recent binary
+     * css version or if an I/O error occurs while reading from the stream
+     */
+    public static Stylesheet loadBinary(InputStream stream) throws IOException {
```
Why do you need to add this public method to the API? I don't see any discussion as to why JavaFX applications need it. It looks like it is just being used internally by `StyleManager`. Unless there is a compelling reason to add this to the API, you will need to make this method package-scope and use an accessor to access it from `StyleManager`.

mstr2 replied:
I don't know why JavaFX applications would need this API. But the argument should be the same argument as for `loadBinary(URL)`, which is also only used internally by `StyleManager`. I think that this isn't opening up any new effective API surface, because it's arguably just another overload of the existing API.

kevinrushforth replied:
An overloaded method with a new signature is still a new method that adds to the public API, although in an easy to understand way. I agree that the same argument could be made for it as for `loadBinary(URL)`, but if neither are needed, then maybe we would be better off to deprecate the existing method (not as part of this enhancement, of course), rather than adding a similar one that also isn't needed as part of the public API. On the other hand, if there is a good reason, I don't see a problem adding this new method.

If you do propose to add this API, you will need to add an appropriate `@since` tag, and you should write an API-level test for this method as part of `StylesheetTest`.

mstr2 replied:
It seems to me that the only useful thing an application developer could do with a `Stylesheet` instance is to inspect its contents (the relevant APIs in javafx.css are public, after all). That seems like a niche use case, but to the extent that there are applications out there that inspect binary stylesheets, having the option to load them from a stream seems at least marginally useful.

I slightly lean towards adding this public API also to not having to introducing a new helper class _just_ for this one benign method. If you have a stronger opinion on this, please go ahead and make the call.

kevinrushforth replied:
Avoiding the need for an accessor isn't sufficient reason, but I do like the symmetry with the existing `loadBinary(URL)` method. Also, one can create a `Stylesheet` by parsing a text file using the various `parse` methods of `CssParser`, so there is ample precedent for doing this. I have no objection to the public method.

kevinrushforth reviewed (commented):
I looked at just the public API and spec, and have two comments:

* I don't see any justification for adding `Stylesheet::loadBinary(InputStream)` to the public API. See my comments inline.
* Do you intend to support setting a user agent stylesheet via a data URL? The docs should be explicit about whether or not a data URI can be used for the following methods:
  * [Application::setUserAgentStylesheet](https://github.com/openjdk/jfx/blob/master/modules/javafx.graphics/src/main/java/javafx/application/Application.java#L521)
  * [Scene::setUserAgentStylesheet](https://github.com/openjdk/jfx/blob/master/modules/javafx.graphics/src/main/java/javafx/scene/Scene.java#L1697)
  * [SubScene::setUserAgentStylesheet](https://github.com/openjdk/jfx/blob/master/modules/javafx.graphics/src/main/java/javafx/scene/SubScene.java#L687)

mstr2 commented:
> * Do you intend to support setting a user agent stylesheet via a data URL? The docs should be explicit about whether or not a data URI can be used for the following methods:
>   
>   * [Application::setUserAgentStylesheet](https://github.com/openjdk/jfx/blob/master/modules/javafx.graphics/src/main/java/javafx/application/Application.java#L521)
>   * [Scene::setUserAgentStylesheet](https://github.com/openjdk/jfx/blob/master/modules/javafx.graphics/src/main/java/javafx/scene/Scene.java#L1697)
>   * [SubScene::setUserAgentStylesheet](https://github.com/openjdk/jfx/blob/master/modules/javafx.graphics/src/main/java/javafx/scene/SubScene.java#L687)

Data URIs also work for user-agent stylesheets. I added documentation to each of the methods.

kevinrushforth commented:
The API and spec changes look good. As I mentioned in a comment added to the CSR, it is ready to move to "Proposed". Once we are farther along in the code review, I'll formally review the CSR.

One thing I noticed is that the diffs don't exactly match what you pasted in the CSR (looks like a copy/paste issue). The `Scene::setUserAgentStylesheet` and `SubScene::setUserAgentStylesheet` methods each have three unchanged lines that are marked with the `+` diff indicator. Can you double-check?

I'll review the implementation next.

mstr2 commented:
> One thing I noticed is that the diffs don't exactly match what you pasted in the CSR (looks like a copy/paste issue). The `Scene::setUserAgentStylesheet` and `SubScene::setUserAgentStylesheet` methods each have three unchanged lines that are marked with the `+` diff indicator. Can you double-check?

I've fixed that, and also cleaned up the `<p>` tags for the new paragraphs.

kevinrushforth commented on modules/javafx.graphics/src/main/java/javafx/css/Stylesheet.java:
```diff
@@ -278,25 +279,48 @@ final void readBinary(int bssVersion, DataInputStream is, String[] strings)
             final int bssVersion = dataInputStream.readShort();
             if (bssVersion > Stylesheet.BINARY_CSS_VERSION) {
-                throw new IOException(url.toString() + " wrong binary CSS version: "
+                throw new IOException("Wrong binary CSS version: "
```
If the `url` string is non-null shouldn't it still be part of the exception message?

mstr2 replied:
True.

kevinrushforth commented on modules/javafx.graphics/src/test/java/test/javafx/css/StylesheetTest.java:
```diff
@@ -646,4 +649,140 @@ public void testRT_37301() {
+        byte[] stylesheetData;
+
+        try {
+            inputFile = File.createTempFile("convertCssTextToBinary", ".css");
+            outputFile = File.createTempFile("convertCssTextToBinary", ".bss");
```
Instead of writing a pair of temp files, maybe you can call `CssParser.parse` to create a `Stylesheet` and then `Stylesheet.writeBinary` to write to an in memory `DataOutputStream`? Or alternatively, create the `.bss file` and add it as a resource under `src/test/resources/...`?

kevinrushforth reviewed (commented):
I spent some time analyzing the changes to `StyleManager::loadStylesheetUnPrivileged`, especially around the use of the local `parse` variable, since you no longer modify it during the processing in a couple places. I think everything is fine, but I'd like @aghaisas to also look at this when he reviews it. If @dsgrieve has any comments, they would be welcome.

The following is the current behavior for the case where `fname` ends with `.css` or `.bss`:

| `fname` | `.css` exists | `.bss` exists | Result |
| ------- | ------------- | ------------- | ------ |
| _SOMENAME_.bss | YES | NO | fail |
| _SOMENAME_.bss | NO | YES | Use _SOMENAME_.bss instead |
| _SOMENAME_.bss | YES | YES | Use _SOMENAME_.bss |
| _SOMENAME_.css | YES | NO | Use _SOMENAME_.css |
| _SOMENAME_.css | NO | YES | Use _SOMENAME_.bss (NOTE: if `-Dbinary.css=false` then fail) |
| _SOMENAME_.css | YES | YES | Use _SOMENAME_.bss (NOTE: if `-Dbinary.css=false` then _SOMENAME_.css) |

In case where `fname` ends with neither `.css` nor `.css`, it is assumed to be a CSS file and parsed as such.

As near as I can tell, both from reading the code and from running tests, this is still working as expected. So there should be no regression in behavior.

The code to process a data URI looks good.

As for the tests, can you add at least one test of calling the public `setUserAgentStylesheet` method using a data URI? Maybe as part of `StyleManagerTest`? Also, I left a couple suggestions inline.

mstr2 commented:
I added a test for each of `Application.setUserAgentStylesheet`, `Scene.setUserAgentStylesheet`, and `SubScene.setUserAgentStylesheet`.

kevinrushforth commented:
Changes look good. I'll do a final pass tomorrow.

aghaisas commented on modules/javafx.graphics/src/test/java/test/com/sun/javafx/css/StyleManagerTest.java:
```diff
@@ -1120,4 +1122,53 @@ public void testCalculateCheckSum() {
+
+            assertEquals(Color.BLUE, rect.getFill());
+        } finally {
+            Application.setUserAgentStylesheet(uaStylesheet);
```
This line in `finally` block results in `ClassNotFoundException`

Run the tests as below to get the log-

`gradle --info :graphics:test &> test.log`

kevinrushforth replied:
Good catch. The reason for this is that the default Stylesheet is in the `javafx.controls` module, which is not available when running unit tests in the `javafx.graphics` module. I recommend to either `@Ignore` this test or else reset it to an empty style sheet (which is effectively what it will be for `javafx.graphics` tests, since the platform initialization code that sets up the default stylesheet is never called for these tests).

mstr2 replied:
I've decided to reset the UA stylesheet to an empty stylesheet.

aghaisas reviewed (commented):
Code changes are OK.
I found an Exception being logged from newly added test. Please check.

kevinrushforth reviewed (approved):
Looks good.

mstr2 commented:
/integrate

kevinrushforth commented:
/sponsor
